lib/Dialect/FIRRTL/Transforms/EliminateWires.cpp

```diff
@@ -0,0 +1,88 @@
+//===- EliminateWires.cpp - Eliminate useless wires -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the EliminateWires pass.  This pass finds wires which,
+// when relocated, become unnecessary and eliminates them.  This is done by
+// creating a node between the write and readers.  Using a node allows the pass
+// to preserve inner symbols and probes, letting  CSE remove useless nodes.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Support/Debug.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/Support/Debug.h"
+
+#include <deque>
+
+#define DEBUG_TYPE "firrtl-eliminate-wires"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct EliminateWiresPass : public EliminateWiresBase<EliminateWiresPass> {
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+// This is the main entrypoint for the lowering pass.
+void EliminateWiresPass::runOnOperation() {
+  LLVM_DEBUG(debugPassHeader(this) << "\n";);
+  auto module = getOperation();
+  auto &dominance = getAnalysis<mlir::DominanceInfo>();
+
+  std::deque<std::pair<WireOp, StrictConnectOp>> worklist;
+
+  for (auto wire : module.getOps<WireOp>()) {
```

Comment:
Is it intentional to walk only wires in the top-level?

Reply:
Yes, other levels should be dead, but even if not, avoiding them simplifies dominance issues.

Reply:
`getSingleConnectUserOf` also only returns connect's from same block, which is important to avoid considering a connect out from a `when` (even if the connect dominates its users) as unconditional.